File: main/friend-circle-queries/friend-circle-queries.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

/** A disjoint-set-union data structure that tracks component sizes. */
final class DisjointSets {
    /**
     * Creates a new singleton set.
     * @return the element index
     */
    int makeSet() {
        _elems.add(-1);
        return _elems.size() - 1;
    }

    /**
     * Unites the sets (union by size).
     * <p>If the elements start out in the same set, no new union happens.</p>
     * @param index1 an element of the first set to participate in the union
     * @param index2 an element of the second set to participate in the union
     * @return the size of the resulting set
     */
    int union(int index1, int index2) {
        // Find the ancestors and stop if they are the same.
        index1 = findSet(index1);
        index2 = findSet(index2);
        if (index1 == index2) return -_elems.get(index1);

        // Unite by size. Sizes are stored negatively, so the index with the
        // more negative associated value will be the parent. If there is a
        // tie, pick index1 as the parent (either could be used in that case).
        return _elems.get(index2) < _elems.get(index1) ? join(index2, index1)
                                                       : join(index1, index2);
    }

    /**
     * Finds the ancestor, performing full path compression.
     * @param index an element of the set whose representative is sought
     * @return the representative element of the set containing {@code index}
     */
    private int findSet(int index) {
        // Find the ancestor.
        int leader = index;
        while (_elems.get(leader) >= 0) leader = _elems.get(leader);

        // Compress the path.
        while (index != leader) {
            final int parent = _elems.get(index);
            _elems.set(index, leader);
            index = parent;
        }

        return leader;
    }

    /**
     * Joins the child to the parent.
     * @param parent a set's representative that will represent the new set
     * @param child a set's representative that will not represent the new set
     * @return the size of the new set
     */
    private int join(final int parent, final int child) {
        _elems.set(parent, _elems.get(parent) + _elems.get(child));
        _elems.set(child, parent);
        return -_elems.get(parent);
    }

    /**
     * When {@code i} is a representative element, {@code _elems[i]} is the
     * negative size of the set containing {@code i}. Otherwise,
     * {@code _elems[i]} is the parent of {@code i}.
     */
    private final List<Integer> _elems = new ArrayList<>();
}

/**
 * Wraps {@code DisjointSets} for an arbitrary key type {@code K}.
 * @param <K> the key type
 */
final class KeyedDisjointSets<K> {
    /**
     * Unites the sets containing {@code key1} and {@code key2}.
     * <p>If a key is not yet an element, a makeset operation is first
     * performed to create a singleton set containing that key.</p>
     * @param key1 an element key of the first set to participate in the union
     * @param key2 an element key of the second set to participate in the union
     * @return the size of the resulting set
     */
    int union(final K key1, final K key2) {
        return _sets.union(index(key1), index(key2));
    }

    /**
     * Obtains or creates the index for the specified key.
     * @param key the key to look up and make a singleton set for if necessary
     * @return the {@code DisjointSets} index corresponding to {@code key}
     */
    private int index(final K key) {
        return _map.computeIfAbsent(key, k -> _sets.makeSet());
    }

    private final Map<K, Integer> _map = new HashMap<>();
    private final DisjointSets _sets = new DisjointSets();
}

final class Solution {
    /**
     * Entry point function. Reads and solves one instance of the problem.
     * @param args Any command-line arguments are ignored.
     */
    public static void main(final String[] args) {
        final KeyedDisjointSets<Integer> ksets = new KeyedDisjointSets<>();
        int maxSize = 0;

        try (final Scanner sc = new Scanner(System.in)) {
            for (int queryCount = sc.nextInt(); queryCount > 0; --queryCount) {
                final int key1 = sc.nextInt(), key2 = sc.nextInt();
                maxSize = Math.max(maxSize, ksets.union(key1, key2));
                System.out.println(maxSize);
            }
        }
    }

    /** The {@code Solution} class should not be instantiated. */
    private Solution() {
        throw new AssertionError();
    }
}

File: main/friend-circle-queries/friend-circle-queries.cpp
#ifdef NDEBUG
#error Assertions unexpectedly disabled.
#endif

#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <unordered_map>
#include <vector>

namespace {
    // A disjoint-set-union data structure that tracks component sizes.
    class DisjointSets {
    public:
        // Creates a new singleton set. Returns the element index.
        int make_set();

        // Union by size. Returns the size of the resulting set.
        int unite(int index1, int index2) noexcept;
        
    private:
        // Finds the ancestor, performing full path compression.
        int find_set(int index) noexcept;

        // Joins the child to the parent. Returns the size of the new set.
        int join(int parent, int child) noexcept;

        // Reports the total number of elements across all sets.
        int count() const noexcept;

        // Reports of an element exists in any set.
        int exists(int index) const noexcept;

        // When i is a representative element, elems_[i] is the negative size
        // of the set containing i. Otherwise, elems_[i] is is the parent of i.
        std::vector<int> elems_;
    };

    int DisjointSets::make_set()
    {
        elems_.push_back(-1);
        return count() - 1;
    }

    int DisjointSets::unite(int index1, int index2) noexcept
    {
        assert(exists(index1));
        assert(exists(index2));

        // Find the ancestors and stop if they are the same.
        index1 = find_set(index1);
        index2 = find_set(index2);
        if (index1 == index2) return -elems_[index1];

        // Union by size. Sizes are stored negatively, so the index with the
        // more negative associated value will be the parent. If there is a
        // tie, pick index1 as the parent (either could be used in that case).
        return elems_[index2] < elems_[index1] ? join(index2, index1)
                                               : join(index1, index2);
    }

    int DisjointSets::find_set(int index) noexcept
    {
        assert(exists(index));

        // Find the ancestor.
        auto leader = index;
        while (elems_[leader] >= 0) leader = elems_[leader];

        // Compress the path.
        while (index != leader) {
            const auto parent = elems_[index];
            elems_[index] = leader;
            index = parent;
        }

        return leader;
    }

    int DisjointSets::join(const int parent, const int child) noexcept
    {
        elems_[parent] += elems_[child];
        elems_[child] = parent;
        return -elems_[parent];
    }

    inline int DisjointSets::count() const noexcept
    {
        return static_cast<int>(elems_.size());
    }

    inline int DisjointSets::exists(const int index) const noexcept
    {
        return 0 <= index && index < count();
    }

    // Wraps DisjointSets for a Key type that is hashable and equatable.
    template<typename Key>
    class KeyedDisjointSets {
    public:
        // Unite the sets containing key1 and key2. Returns the size of that
        // set. If a key is not yet an element, a makeset operation is first
        // performed to create a singleton set containing that key.
        int unite(Key key1, Key key2);

    private:
        // Obtains or creates the index for the specified key.
        int index(Key key);

        std::unordered_map<Key, int> map_;
        DisjointSets sets_;
    };

    template<typename Key>
    int KeyedDisjointSets<Key>::unite(const Key key1, const Key key2)
    {
        return sets_.unite(index(key1), index(key2));
    }

    template<typename Key>
    int KeyedDisjointSets<Key>::index(const Key key)
    {
        const auto it = map_.find(key);
        if (it != end(map_)) return it->second;

        const auto ret = sets_.make_set();
        map_.emplace(key, ret);
        return ret;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    int query_count {};
    std::cin >> query_count;
    assert(query_count > 0);

    KeyedDisjointSets<int> ksets;
    auto max_size = 0;

    while (query_count-- > 0) {
        int key1 {}, key2 {};
        std::cin >> key1 >> key2;
        assert(key1 > 0 && key2 > 0);
        
        max_size = std::max(max_size, ksets.unite(key1, key2));
        std::cout << max_size << '\n';
    }
}
